### PR TITLE
Make `std` an optional dependency. Require `alloc` instead.

### PR DESCRIPTION
**Pull request by briansmith**
This raises the minimum version of Rust required to 1.36.

### CODE REVIEW

**Comment by josephlr** [comment]
This is great! If you're making backwards-incompatible changes to the Cargo features, it might be worth reconsidering the `dev_urandom_fallback` feature.

If I'm understanding the motivation correctly, the main reason to not have the `dev_urandom_fallback` feature is to avoid depending on `libstd`. However, if we now have separate `std` and `alloc` features, it might make sense to have the `std` feature control if we read from `/dev/urandom` on Linux.

Once #558 is fixed, there's not a good _security_ reason to disable the fallback. However, this might be better addressed in a followup PR.

File: src/error.rs [hunk 1]
```diff
@@ -36,7 +36,7 @@ use untrusted;
 /// enum Error {
 ///     CryptoError,
 ///
-/// #  #[cfg(feature = "use_heap")]
+/// #  #[cfg(feature = "alloc")]
```
Comment (josephlr) [inline src/error.rs]: `feature = "std"`

File: src/test.rs [hunk 2]
```diff
@@ -116,13 +116,13 @@
 //! stack trace to the line in the test code that panicked: entry 9 in the
 //! stack trace pointing to line 652 of the file `example.rs`.
 
-#[cfg(feature = "use_heap")]
-use crate::bits;
+use crate::{bits, digest, error};
```
Comment (josephlr) [inline src/test.rs]: Should the `bits` module be gated on the `alloc` feature here? Like it is in `lib.rs`?

File: src/lib.rs [hunk 3]
```diff
@@ -62,13 +65,15 @@
...
+extern crate alloc;
+
+#[cfg(any(test, feature = "std"))]
 extern crate std;
 
 #[macro_use]
 mod debug;
 
-#[cfg(any(test, feature = "use_heap"))]
 #[macro_use]
```
Comment (josephlr) [inline src/lib.rs]: This still needs a `#[cfg(test)]` in order for `cargo build --no-default-features` to work.

Comment (josephlr) [reply src/lib.rs]: Actually, `#[cfg(feature = "alloc")]` is what you want here, so that the integration tests keep working.

File: src/lib.rs [hunk 4]
```diff
@@ -62,13 +65,15 @@
 #![no_std]
 #![cfg_attr(feature = "internal_benches", allow(unstable_features), feature(test))]
 
-#[cfg(any(test, feature = "use_heap"))]
+#[cfg(any(test, feature = "alloc"))]
```
Comment (josephlr) [inline src/lib.rs]: `#[cfg(feature = "alloc")]` is all that's needed here, as `test` requires `alloc`.

In general, there a lot of `#[cfg(any(test, feature = "alloc"))]` in the code. These should be simplified to `#[cfg( feature = "alloc")]

File: src/lib.rs [hunk 5]
```diff
@@ -62,13 +65,15 @@
...
+extern crate alloc;
+
+#[cfg(any(test, feature = "std"))]
```
Comment (josephlr) [inline src/lib.rs]: `#[cfg(feature = "std")]` is what's needed here, otherwise the current `cfg` make tests require `std`.

File: src/test.rs [hunk 6]
```diff
@@ -116,13 +116,13 @@
...
 
-use crate::{digest, error};
+#[cfg(feature = "alloc")]
```
Comment (josephlr) [inline src/test.rs]: If `#[cfg(feature = "alloc")]` is added to `lib.rs` then these annotations are not needed in this file (here and elsewhere).

**Comment by briansmith** [comment]
Thanks. I updated the PR to address most of your concerns.

There is more work that needs to be done to properly support testing for `no_std` environments, e.g. adding the non-default configurations to CI. Those changes will have to wait until my pipeline is cleared.

**Comment by briansmith** [comment]
Thanks for the review!

**Pull request merged**